### DISCUSSION TITLE
Async Code in the Dynamic Sitemap Function

### DISCUSSION THREAD

**Discussion by NeeshSamsi** [vercel/next.js#50787] [App Router] [answered]
I'm using the sitemap.ts file in the app router to generate a dynamic sitemap. The data I am pulling to display on the sitemap is using an async function but as soon as I add async to the sitemap() function, I get the following error: 

Type 'MetadataRoute.Sitemap' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.

```ts
// sitemap.ts

import { MetadataRoute } from "next"
import * as config from "@/lib/config"
import { getPosts } from "@/lib/posts"
import { getCategories } from "@/lib/categories"

export default async function sitemap(): MetadataRoute.Sitemap {
  const pages = ["", "blog", "categories"]
  const pageUrls = pages.map((page) => ({
    url: `${config.url}/${page}`,
    lastModified: new Date(),
  }))

  const posts = await getPosts()
  const postUrls =
    posts &&
    posts.map((post) => ({
      url: `${config.url}/posts/${post.id}`,
      lastModified: new Date(post.date),
    }))

  const categories = getCategories()
  const categoryUrls = categories.map((tag) => ({
    url: `${config.url}/blog/categories/${tag}`,
    lastModified: new Date(),
  }))

  return [...pageUrls, ...postUrls, ...categoryUrls]
}
```
Here the only issue is the getPosts function which needs to asynchronously parse posts.

Is there a way that the sitemap.ts file supports async functions? Or is there some kind of workaround I can use? Or do I have to have synchronous code in the dynamic sitemap function?

**Comment by icyJoseph** [comment, marked as answer]
If you are going to type explicitly, you gotta wrap it with the Promise generic object type, `Promise<MetadataRoute.Sitemap>`:

```ts
// sitemap.ts
import type { MetadataRoute } from "next";

export default async function sitemap(): Promise<MetadataRoute.Sitemap> {
  const posts = await Promise.resolve().then(() => [
    { id: "0", date: "2023-05-06" },
  ]);
  const postUrls =
    posts &&
    posts.map((post) => ({
      url: `http://example.com/posts/${post.id}`,
      lastModified: new Date(post.date),
    }));

  return [...postUrls];
}

```

**Reply by NeeshSamsi** [reply]
Typing it with a Promis<> satisfies the Type error I was getting. But now I am getting the following error:

`Error occurred prerendering page "/sitemap.xml". Read more: https://nextjs.org/docs/messages/prerender-error
TypeError: Cannot read properties of undefined (reading 'handle')`

The complete code is as follows: 

```ts
// sitemap.ts
import { MetadataRoute } from "next"
import * as config from "@/lib/config"
import { getPosts } from "@/lib/posts"
import { getCategories } from "@/lib/categories"

export default async function sitemap(): Promise<MetadataRoute.Sitemap> {
  const pages = ["", "blog", "categories"]
  const pageUrls = pages.map((page) => ({
    url: `${config.url}/${page}`,
    lastModified: new Date(),
  }))

  const posts = await getPosts()
  const postUrls = posts.map((post) => ({
    url: `${config.url}/blog/${post.id}`,
    lastModified: new Date(post.date),
  }))

  const categories = getCategories()
  const categoryUrls = categories.map((tag) => ({
    url: `${config.url}/blog/categories/${tag.slug}`,
    lastModified: new Date(),
  }))

  return [...pageUrls, ...postUrls, ...categoryUrls]
}
```

```ts
// posts.tsx
// Mostly unrelated but here it is
export async function getPosts(
  noOfPosts?: number,
  categorySlug?: string
): Promise<BlogPostMetaParsed[]> {
  // Get blog filenames
  const fileNames = fs.readdirSync(postsDir)

  const allPostsData = await Promise.all(
    fileNames.map(async (fileName) => {
      // Remove .mdx from name to get id
      const id = fileName.replace(/\.mdx$/, "")

      const { blogPost } = await parsePost(id, fileName)

      return blogPost
    })
  )

  let limit

  if (noOfPosts !== -1) {
    limit = noOfPosts
  }

  // Sort by date
  const sortedPosts = allPostsData.sort((a, b) => (a.date < b.date ? 1 : -1))

  let allPosts = sortedPosts

  // If category is requested check category exists
  if (categorySlug) {
    const category = getCategory(categorySlug)

    if (!category) {
      return []
    } else {
      allPosts = sortedPosts.filter((post) => {
        return post.tags.some((tag) => tag.slug === categorySlug)
      })
    }
  }

  // If limit is requested, slice to that size
  allPosts = limit ? allPosts.slice(0, limit) : allPosts

  return allPosts
}
async function parsePost(id: string, fileName: string) {
  // Read md file as string
  const fullPath = path.join(postsDir, fileName)
  const fileContents = fs.readFileSync(fullPath, "utf8")

  const { frontmatter, content } = await compileMDX<BlogPostMeta>({
    source: fileContents,
    options: {
      parseFrontmatter: true,
      // mdxOptions: {
      //   rehypePlugins: [
      //     rehypeSlug,
      //     [rehypeAutolinkHeadings, { behavior: "wrap" }],
      //   ],
      // },
    },
  })

  const tags: Category[] = frontmatter.tags.map((tag: string) => {
    const category = getCategory(tag)
    if (!category) {
      throw new Error(`Invalid tag in post: ${id}`)
    }
    return category
  })

  const authors: Author[] = frontmatter.authors.map((authorSlug: string) => {
    const author = getAuthor(authorSlug)
    if (!author) {
      throw new Error(`Invalid author in post: ${id}`)
    }
    return author
  })

  const blogPost: BlogPostMetaParsed = {
    id,
    title: frontmatter.title,
    description: frontmatter.description,
    tags,
    imgUrl: frontmatter.imgUrl,
    imgAlt: frontmatter.imgAlt,
    authors,
    date: frontmatter.date,
  }

  return { blogPost, content }
}
```

**Reply by icyJoseph** [reply]
Well looks like that has to be another discussion. 

I'd say, remove posts and categories, that is comment all of their code, and try to build. If that works, then we know pages is ok. So now we introduce back posts, and so on... 

Also since sitemap is doing async work, you could benefit from wrapping it all in a try/catch and analyse what you capture in the catch block.

**Reply by NeeshSamsi** [reply]
I've already verified that categories and pages work, the xml file renders perfectly:

https://inquisit.vercel.app/sitemap.xml
```xml
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
<url>
<loc>https://inquisit.vercel.app/</loc>
<lastmod>2023-06-05T15:03:55.602Z</lastmod>
</url>
<url>
<loc>https://inquisit.vercel.app/blog</loc>
<lastmod>2023-06-05T15:03:55.602Z</lastmod>
</url>
<url>
<loc>https://inquisit.vercel.app/categories</loc>
<lastmod>2023-06-05T15:03:55.602Z</lastmod>
</url>
<url>
<loc>https://inquisit.vercel.app/blog/categories/cognitive-science</loc>
<lastmod>2023-06-05T15:03:55.602Z</lastmod>
</url>
<url>
<loc>https://inquisit.vercel.app/blog/categories/evolutionary-biology</loc>
<lastmod>2023-06-05T15:03:55.602Z</lastmod>
</url>
</urlset>
```

I tried using a trycatch and also just tried using a console.log but I get no console output. It seems that it doesn't even run before catching some sort of error. This is why I am wondering whether the sitemap() function somehow doesn't allow for async code to be run because, before I was using a different implementation to parse posts and that was not async. The sitemap worked flawlessly, it's introducing async causes erorrs.

**Reply by icyJoseph** [reply]
Does my example with Promise resolve work? That'd be equivalent. I remember doing a call to that endpoint and getting XML back.

**Reply by icyJoseph** [reply]
I wonder if the problem is that a post doesn't exist at all when trying to open the file handle.

**Reply by NeeshSamsi** [reply]
Oh no way, it's working with the simulating getting posts. But I am using the same getPosts function to render posts in the main blog, I wonder why it's only causing errors in this file. As for a post not existing, there's currently only a single post and again if it works in the main app, I wonder why it would cause errors in the sitemap.

I don't understand what a file handle is.

I really appreciate you helping out. If you would like to take a closer look, here's the repo: https://github.com/NeeshSamsi/inquisit. However I totally understand if you don't want to sink in more time into looking into this.

**Reply by icyJoseph** [reply]
Will check when I'm home 👍

**Reply by icyJoseph** [reply]
You know, with your project setup, which uses turbopack in dev mode always responded with 404.

So, anyway, I tried without turbopack and it failed.

```console
error TypeError: Cannot read properties of undefined (reading 'setup')
    at RouteHandlerManager.handle 
```

Next I tried building instead:

```console
[    ] - info Generating static pages (0/11)
Error occurred prerendering page "/sitemap.xml". Read more: https://nextjs.org/docs/messages/prerender-error
TypeError: Cannot read properties of undefined (reading 'handle')
- info Generating static pages (11/11)
```

All of this smells like a problem with Next.js rather, so I check what version you have, 13.1.4 - So I upgraded to 13.4.4

When I built for prod this, with 13.4.4, using this file:

```ts
import type { MetadataRoute } from "next"
import * as config from "@/lib/config"
import { getPostIDs, getPosts } from "@/lib/posts"
import { getCategories } from "@/lib/categories"

export default async function sitemap(): Promise<MetadataRoute.Sitemap> {
  const pages = ["", "blog", "categories"]
  const pageUrls = pages.map((page) => ({
    url: `${config.url}/${page}`,
    lastModified: new Date(),
  }))

  const posts = await getPosts()
  const postUrls = posts.map((post) => ({
    url: `${config.url}/blog/${post.id}`,
    lastModified: new Date(post.date),
  }))

  const categories = getCategories()
  const categoryUrls = categories.map((tag) => ({
    url: `${config.url}/blog/categories/${tag.slug}`,
    lastModified: new Date(),
  }))

  return [...pageUrls, ...postUrls, ...categoryUrls]
}

```

I get:

```xml

<?xml version="1.0" encoding="UTF-8"?>
<urlset 
  xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <url>
    <loc>https://inquisit.vercel.app/</loc>
    <lastmod>2023-06-05T20:43:01.684Z</lastmod>
  </url>
  <url>
    <loc>https://inquisit.vercel.app/blog</loc>
    <lastmod>2023-06-05T20:43:01.684Z</lastmod>
  </url>
  <url>
    <loc>https://inquisit.vercel.app/categories</loc>
    <lastmod>2023-06-05T20:43:01.684Z</lastmod>
  </url>
  <url>
    <loc>https://inquisit.vercel.app/blog/the-emergence-of-intelligence</loc>
    <lastmod>2023-05-23T22:00:00.000Z</lastmod>
  </url>
  <url>
    <loc>https://inquisit.vercel.app/blog/categories/cognitive-science</loc>
    <lastmod>2023-06-05T20:43:01.687Z</lastmod>
  </url>
  <url>
    <loc>https://inquisit.vercel.app/blog/categories/evolutionary-biology</loc>
    <lastmod>2023-06-05T20:43:01.687Z</lastmod>
  </url>
</urlset>
```

Though, dev mode with turbopack still responds with 404.


https://github.com/vercel/next.js/assets/21013447/ee34008e-1a27-4fb3-81b5-5e3df7722545

The package.json in the video has dev mode using turbopack, but I started the server with `npx next dev`.

## TLDR

Upgrade to latest next.js - 13.4.4 at least, and maybe avoid turbopack just now.
